### PR TITLE
Feature/can integration

### PR DESCRIPTION
**Pull request by dejfcz**
Protocol related to CAN integration into PS + support for client calls related to CAN.

@hellwolf @wigy-opensource-developer @Luis-Fernando-Molina

### CODE REVIEW

File: tests/PS09.md
Comment (wigy-opensource-developer) [inline tests/PS09.md:97]: I am a bit confused here. Does the client need to do any actual transformation from binary to base58 here? If it would be possible, I would leave it to the CAN to both calculate the hash of the content and to find a suitable encoding/serialization format for that binary hash. Ideally, this line should be just `$ObjectPath1 := "/ipfs/" + $ObjectHash1` or even `$ObjectPath1 := $ObjectHash1`

File: tests/PS09.md
Comment (wigy-opensource-developer) [inline tests/PS09.md:181]: I was confused first, because I thought TTL was set to 60 seconds in step 5, not 6 seconds. Maybe a comment would help there, because the number of zeros is too high for humans.

File: tests/PS09.md
Comment (wigy-opensource-developer) [inline tests/PS09.md]: What should be the hash returned there?

File: tests/PS09.md
Comment (wigy-opensource-developer) [inline tests/PS09.md:206]: A normal client hopefully does not have access to the admin interface of the CAN. Why is it accessed by the test here?

File: tests/PS09.md
Comment (wigy-opensource-developer) [inline tests/PS09.md:396]: Is it some hash value that cannot be parsed by IPFS or is it something you just made up, so it is unlikely to exist on the network?

**Review by wigy-opensource-developer — COMMENTED** [review]
The only critical comment is about the hash encoding. Otherwise I like the protocol and acceptance tests.

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md:181]: OK OK

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md:97]: Currently, this format of value in IPNS record is the only one that is supported. I'm not even sure if IPFS would accept anything else. If IPFS would accept it, it is possible to extend this. To me it seemed like not a big deal for the client to construct the path like this, so I did not even checked what other formats are supported.

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md]: Will fix, tx!

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md:206]: This is test of profile server functionality, I want to verify that PS did its job.

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md:396]: It is a valid hash just to make sure the PS won't report any problem because of the format of the value. We are testing something else here, namely we test that profile server will not allow the client to setup IPNS to any other CAN object that it currently owns. And here, it owns no objects, so we require an error here from PS.

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md:206]: Otherwise you are right, the client will not have access to admin interface.

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md:97]: The best thing to do here is probably to explicitly define this in the protocol comments. IPFS is confusing here and I'm not sure if it is a bug or feature. Thing is that those object hashes do have multihash format, but do not have multibase format. So I would expect to see two kinds of prefixes there - one to say it is SHA256 hash and 32 bytes of that and the second saying that it is base58 encoding - and this second is not being used here. Therefore it is probably best to just say what should be there, otherwise it is really confusing (and took me a lot of time to understand while coding actually).

Comment (dejfcz) [reply tests/PS09.md:97]: Should be now fixed in the latest commit.

File: tests/PS09.md
Comment (wigy-opensource-developer) [reply tests/PS09.md:396]: So I cannot set my profile to the latest Madonna album. Great.

File: tests/PS09.md
Comment (wigy-opensource-developer) [reply tests/PS09.md:97]: Yeah, not having the base58 prefix is already historical in a project having so short history :smile: .

My actual concern is that CanStoreDataResponse.hash should already return something like "QmaRvpLT4RgBDVY6cQKwqQGQNtbM31KSqh8TxvQgFoEUus" or "/ipfs/QmaRvpLT4RgBDVY6cQKwqQGQNtbM31KSqh8TxvQgFoEUus" in my opinion. That futureproofs this API in case IPFS starts using base64 or some other serialization for the content hash.

Comment (dejfcz) [reply tests/PS09.md:97]: It returns just QmaRvpLT4RgBDVY6cQKwqQGQNtbM31KSqh8TxvQgFoEUus. If there is an update in future that will replace this weird behavior, we can review this on our side as well. Just worked with the version I have, so good for me at this point. There should be no problem if the default encoding was switched to base64. As long as base58 is supported, we should be fine.

Comment (wigy-opensource-developer) [reply tests/PS09.md:97]: Good. If client can just take CanStoreDataResponse.hash, prefix it with "/ipfs/" and put it into CanPublishIpnsRecordRequest.record.value, then I am fine. No base58 transformation is needed then in the client.

**Review by wigy-opensource-developer — APPROVED** [review]

File: tests/PS09.md
Comment (dejfcz) [reply tests/PS09.md:97]: Sorry, misunderstanding. PS will return the binary form of a multihash. The client currently needs to use base58 encoding on that and prefix that with "/ipfs/".

**Review by wigy-opensource-developer — CHANGES_REQUESTED** [review]
I think the protocol would be more future proof if CanStoreDataRequest.hash would contain the content hash in the form returned by the CAN node. Assuming a given base encoding, decoding it in PS and asking the client to encode it back seems to be a waste of resources and increases coupling.

**Comment by dejfcz** [comment]
I think that giving the client the very raw multihash is much better is it is free to use it without additional parsing against the IPFS gateway. There should be no problem for the client to encode it for the purpose of creating IPNS. I would rather work with objects in their raw form, rather than creating BLOBs that you should not touch.

It is currently impossible to avoid the implementation of the encoding on the client as it is required when dealing with IPNS resolving of other identity. So the client will have to understand and implement the encoding anyway. If we remove use of encoding in the construction of IPNS, it will just removes it from there and not elsewhere.

**Comment by dejfcz** [comment]
Seems that we came to a temporary resolution here - to leave it as it is and watch how IPFS evolves as it seems too unpredictable at this moment to justify that any way is better than the other.

**Comment by dejfcz** [comment]
Does not seem like more reviews coming ... will merge this in couple of days unless something new comes.

**Pull request merged**
